libcontainer/capabilities/capabilities.go

```diff
@@ -34,50 +35,58 @@ func init() {
 }
 
 // New creates a new Caps from the given Capabilities config. Unknown Capabilities
-// or Capabilities that are unavailable in the current environment produce an error.
+// or Capabilities that are unavailable in the current environment are ignored,
+// printing a warning instead.
 func New(capConfig *configs.Capabilities) (*Caps, error) {
 	var (
-		err error
-		c   Caps
+		err      error
+		c        Caps
+		warnings = make(map[string]struct{})
 	)
 
-	if c.caps[capability.BOUNDING], err = capSlice(capConfig.Bounding); err != nil {
-		return nil, err
-	}
-	if c.caps[capability.EFFECTIVE], err = capSlice(capConfig.Effective); err != nil {
-		return nil, err
-	}
-	if c.caps[capability.INHERITABLE], err = capSlice(capConfig.Inheritable); err != nil {
-		return nil, err
-	}
-	if c.caps[capability.PERMITTED], err = capSlice(capConfig.Permitted); err != nil {
-		return nil, err
-	}
-	if c.caps[capability.AMBIENT], err = capSlice(capConfig.Ambient); err != nil {
-		return nil, err
+	c.caps = map[capability.CapType][]capability.Cap{
+		capability.BOUNDING:    capSlice(capConfig.Bounding, warnings),
+		capability.EFFECTIVE:   capSlice(capConfig.Effective, warnings),
+		capability.INHERITABLE: capSlice(capConfig.Inheritable, warnings),
+		capability.PERMITTED:   capSlice(capConfig.Permitted, warnings),
+		capability.AMBIENT:     capSlice(capConfig.Ambient, warnings),
 	}
 	if c.pid, err = capability.NewPid2(0); err != nil {
 		return nil, err
 	}
 	if err = c.pid.Load(); err != nil {
 		return nil, err
 	}
+	printWarnings(warnings)
 	return &c, nil
 }
 
-// capSlice converts the slice of capability names in caps, to their numeric
+// capSlice converts the slice of capability names in newCaps, to their numeric
 // equivalent, and returns them as a slice. Unknown or unavailable capabilities
-// produce an error.
-func capSlice(caps []string) ([]capability.Cap, error) {
-	out := make([]capability.Cap, len(caps))
-	for i, c := range caps {
-		v, ok := capabilityMap[c]
-		if !ok {
-			return nil, fmt.Errorf("unknown capability %q", c)
+// are not returned, but appended to warnings, which can be used by the caller to
```

Comment:
I'd say "but added to warnings map" and stop here (it's up to the caller what to do with them -- print a warning, an error, or set the world on fire).